File: 5.rb
require "minitest/autorun"

# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

class ProjectEulerTests < Minitest::Test
  def least_common_multiple(max)
    lcm = 1
    (1..max).each { |i| lcm *= (i / gcd(lcm, i)) }

    lcm
  end

  def gcd(a, b)
    while b > 0
      a %= b
      return b if a == 0
      b %= a
    end

    a
  end

  def test_least_common_multiple
    assert_equal 2520, least_common_multiple(10)
    assert_equal 232792560, least_common_multiple(20)
  end
end

File: 5.exs
defmodule SmallestMultiple do
@moduledoc """

  Problem 5: Smallest multiple

  2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
  What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

  """
  def smallest(1), do: 1
  def smallest(2), do: 2

  def smallest(n) when n > 2 do
    next = smallest(n-1)
    case rem(next, n) do
      0 -> next
      _ -> next * div(n,gcd(next,n))
    end
  end

  def gcd(x, 0), do: x
  def gcd(x, y), do: gcd(y, rem(x,y))
end


IO.puts SmallestMultiple.smallest(20)
